File: Preferences/Preferences/ColorToValueTransformer.m
//
//  ColorToValueTransformer.m
//  Preferences
//
//  Created by Christopher Snowhill on 5/22/22.
//

#import "ColorToValueTransformer.h"

@implementation ColorToValueTransformer

+ (Class)transformedValueClass {
	return [NSColor class];
}
+ (BOOL)allowsReverseTransformation {
	return YES;
}

// Convert from string to NSURL
- (id)reverseTransformedValue:(id)value {
	if(value == nil) return nil;

	NSError *error;
	NSData *data = [NSKeyedArchiver archivedDataWithRootObject:value
	                                     requiringSecureCoding:YES
	                                                     error:&error];

	return data;
}

- (id)transformedValue:(id)value {
	if(value == nil) return [NSColor colorWithRed:0 green:0 blue:0 alpha:1.0];

	NSError *error;
	NSColor *color = nil;

	if(@available(macOS 11.0, *)) {
		color = (NSColor *)[NSKeyedUnarchiver unarchivedArrayOfObjectsOfClass:[NSColor class]
		                                                             fromData:value
		                                                                error:&error];
	} else {
		NSSet *allowed = [NSSet setWithArray:@[[NSColor class]]];
		color = (NSColor *)[NSKeyedUnarchiver unarchivedObjectOfClasses:allowed
		                                                       fromData:value
		                                                          error:&error];
	}

	return color;
}

@end

File: Transformers/ColorToValueTransformer.swift
import Cocoa

@objc(ColorToValueTransformer)
class ColorToValueTransformer: ValueTransformer {
    override class func transformedValueClass() -> AnyClass {
        return NSColor.self
    }

    override class func allowsReverseTransformation() -> Bool {
        return true
    }

    override func transformedValue(_ value: Any?) -> Any? {
        guard let value = value else {
            return NSColor(red: 0, green: 0, blue: 0, alpha: 1.0)
        }

        if let value = value as? NSData {
            do {
                let unarchiver = try NSKeyedUnarchiver(forReadingFrom: value as Data)
                unarchiver.requiresSecureCoding = true
                let classes = [NSColor.self]
                let ret = unarchiver.decodeObject(of: classes, forKey: "root")
                return ret ?? NSColor(red: 0, green: 0, blue: 0, alpha: 1.0)
            } catch {
                NSLog("Error unarchiving color: \(error)")
                return NSColor(red: 0, green: 0, blue: 0, alpha: 1.0)
            }
        }

        return NSColor(red: 0, green: 0, blue: 0, alpha: 1.0)
    }

    override func reverseTransformedValue(_ value: Any?) -> Any? {
        guard let value = value else {
            return nil
        }

        let data = try? NSKeyedArchiver.archivedData(withRootObject: value, requiringSecureCoding: true)

        return data
    }
}
